core/src/main/scala/io/prediction/core/SelfCleaningDataSource.scala

```diff
@@ -0,0 +1,313 @@
+    Await.result(futureOfListNewEvents, scala.concurrent.duration.Duration(60, "minutes")) 
+
+    removeEvents(eventsToRemove, appId)
+  }
+
+
+  /** :: DeveloperApi ::
+    *
+    * Filters most recent, compress properties of PEvents
+    */
+  @DeveloperApi
+  def cleanPEvents(sc: SparkContext): RDD[Event] = {
+   val pEvents = PEventStore.find(appName)(sc).sortBy(_.eventTime)
+
+   val rdd = eventWindow match {
+      case Some(ew) =>
+        var updated =
+          if (ew.compressProperties) compressPProperties(sc, pEvents) else pEvents
+        
+        val deduped = if (ew.removeDuplicates) removePDuplicates(sc,updated) else updated
+        deduped
+      case None =>
+        pEvents
+    }
+  getCleanedPEvents(rdd)
+  }
+
+  /** :: DeveloperApi ::
+    *
+    * Filters most recent, compress properties and removes duplicates of LEvents
+    *
+    * @return Iterator[Event] most recent LEvents
+    */
+  @DeveloperApi
+  def cleanPersistedLEvents: Unit = {
+    eventWindow match {
+      case Some(ew) =>
+
+        val result = cleanLEvents().toSet
+        val originalEvents = LEventStore.find(appName).toSet
+        val newEvents = result -- originalEvents
+        val eventsToRemove = (originalEvents -- result).map { case e =>
+          e.eventId.getOrElse("")
+        }
+
+        wipe(newEvents, eventsToRemove)
+  
+       case None =>
+    }
+  }
+
+  /** :: DeveloperApi ::
+    *
+    * Filters most recent, compress properties of LEvents
+    */
+  @DeveloperApi
+  def cleanLEvents(): Iterable[Event] = {
+    val lEvents = LEventStore.find(appName).toList.sortBy(_.eventTime)
+ 
+    val events = eventWindow match {
+      case Some(ew) =>
+        var updated =
+          if (ew.compressProperties) compressLProperties(lEvents) else lEvents
+          val deduped = if (ew.removeDuplicates) removeLDuplicates(updated) else updated
+        deduped
+      case None =>
+        lEvents
+    }
+    getCleanedLEvents(events)
+  }
+
+
+  private def isSetEvent(e: Event): Boolean = {
+    e.event == "$set" || e.event == "$unset"
+  }
+
+  private def compress(events: Iterable[Event]): Event = {
+    events.find(_.event == "$set") match {
+
+      case Some(first) =>
+        events.reduce { (e1, e2) =>
+          val props = e2.event match {
+            case "$set" =>
+              e1.properties.fields ++ e2.properties.fields
+            case "$unset" =>
+              e1.properties.fields
+                .filterKeys(f => !e2.properties.fields.contains(f))
+          }
+          e1.copy(properties = DataMap(props))
+        }
+
+      case None =>
+        events.reduce { (e1, e2) =>
+          e1.copy(properties =
+            DataMap(e1.properties.fields ++ e2.properties.fields)
+          )
+        }
+    }
+  }
+}
+
+case class EventWindow(
+  duration: Option[String] = None,
+  removeDuplicates: Boolean = false,
+  compressProperties: Boolean = false
+)
```

Comment:
Add the notion of an EventWindow to the Datasource. This provides for deduping events, compressing $set/$unset events, and trimming older events at the desired watermark.